can: isotp: Add CAN-FD support to ISO-TP

gramsay0 opened the pull request:
- Add CAN-FD support to ISO-TP
- Update ISO-TP sample to work with CAN-FD
- Update existing ISO-TP unit tests to work with CAN-FD
- Add CAN-FD specific unit tests
  - Mandatory padding of frames > 8 bytes to next DLC
  - Validation that RX_DL in consecutive frames matches first frame

Thanks @dale-herman for doing most of the work

gramsay0 committed:
canbus: isotp: add CAN-FD support

ISO-TP CAN-FD support can be enabled at runtime.

Signed-off-by: Grant Ramsay <gramsay@enphaseenergy.com>
Signed-off-by: Martin Jäger <martin@libre.solar>

gramsay0 committed:
samples: canbus: isotp: Add support for CAN-FD

Enable this sample to be used with CAN-FD

Signed-off-by: Grant Ramsay <gramsay@enphaseenergy.com>

alexanderwachter commented:
Just scrolled over it. Looks good at a first glance! You may extend the copyright notice, because it is a bigger extension of functionality.

gramsay0 commented:
> Just scrolled over it. Looks good at a first glance! You may extend the copyright notice, because it is a bigger extension of functionality.

@alexanderwachter thanks!

If you have time, could you please have a look over this RFC https://github.com/zephyrproject-rtos/zephyr/issues/59896 and give some feedback?
Apologies for not tagging you in it sooner, I would appreciate your input

gramsay0 commented on tests/subsys/canbus/isotp/conformance/testcase.yaml:
```diff
@@ -5,3 +5,36 @@ tests:
+      - CONFIG_ISOTP_TX_DL_MAX=8
+    platform_allow:
+      - native_posix
+      - native_posix_64
```
Is there a twister filter like `dt_chosen_enabled("zephyr,canbus")` for CAN-FD?
For now, I only enabled ISO-TP CAN-FD testing on native posix

henrikbrixandersen replied:
No, you cannot filter tests based on CAN-FD support. You'll need to skip tests if runtime CAN-FD support is not available.

gramsay0 replied:
Good idea, I have made that change

martinjaeger reviewed (changes requested):
Thanks for the contribution, only got some minor comments.

May I ask you to move the changes related to `CONFIG_ISOTP_ENABLE_TX_PADDING` and `CONFIG_ISOTP_REQUIRE_RX_PADDING` into a dedicated commit? As far as I can see they are not related to CAN-FD support. This will make it easier to understand why the change was made in the future.

gramsay0 commented on samples/subsys/canbus/isotp/Kconfig:
```diff
@@ -10,4 +10,19 @@ config SAMPLE_LOOPBACK_MODE
+# A frame has 7 bytes payload, we are using a BS of 8 and need one char for the
+# string termination (7 * 8 + 1 = 57)
+config ISOTP_RX_BUF_SIZE
+	default 63 if SAMPLE_CAN_FD_MODE
```
I think this can be simplified by using format precision to print the non-null terminated strings:

    printk("%.*s", buf->len, buf->data);

Although, I think some of the minimal string formatting libraries don't support precision.
`k_str_out` seems to be a good alternative:

    k_str_out(buf->data, buf->len);

martinjaeger replied:
As far as I can see we don't even need it as the terminating null character is also sent over ISO-TP (using `sizeof` and not `strlen` in the send function).

gramsay0 replied:
It's the receive side that is the issue (`rx_8_0_thread()`).
The packet is returned in partial chunks and printed.
The partial chunks may not be null terminated, but the length is known

martinjaeger replied:
Ok. I see. `k_str_out` is marked as an internal API, so I'd just go with the `printk("%.*s"...)`  if it works with the default configuration of this sample.

gramsay0 commented on subsys/canbus/isotp/isotp.c:
```diff
@@ -137,13 +139,13 @@ static void receive_send_fc(struct isotp_recv_ctx *ctx, uint8_t fs)
 	*data++ = ctx->opts.stmin;
 	payload_len = data - frame.data;
 
-#if defined(CONFIG_ISOTP_REQUIRE_RX_PADDING) || \
-				defined(CONFIG_ISOTP_ENABLE_TX_PADDING)
+#if defined(CONFIG_ISOTP_ENABLE_TX_PADDING)
```
My understanding is:

- `ISOTP_ENABLE_TX_PADDING` makes the device AUTOSAR conformant, by sending TX padding
- `ISOTP_REQUIRE_RX_PADDING` ensures the other devices on the bus are AUTOSAR conformant, by rejecting non-padded frames

In this case the device is transmitting a frame, so `ISOTP_ENABLE_TX_PADDING` alone would enable TX padding

I don't see a reason for one config to depend on the other

martinjaeger replied:
Yes, that's correct.

But does it make sense to enforce padding for other bus participants if you don't use it yourself?

No strong opinions on that, we can also keep it as is. But the padding fix should still go into a dedicated commit as it has nothing to do with CAN-FD.

gramsay0 replied:
> But does it make sense to enforce padding for other bus participants if you don't use it yourself?

I agree that would be a strange case...
Maybe `default y if ISOTP_ENABLE_TX_PADDING` would be a good approach, as it is more of a suggestion than a hard dependency.
Yes, I will move to a separate PR

martinjaeger replied:
That sounds like the best approach!

gramsay0 commented:
> May I ask you to move the changes related to `CONFIG_ISOTP_ENABLE_TX_PADDING` and `CONFIG_ISOTP_REQUIRE_RX_PADDING` into a dedicated commit? As far as I can see they are not related to CAN-FD support. This will make it easier to understand why the change was made in the future.

@martinjaeger Okay, will do

gramsay0 commented:
@martinjaeger here is the padding fix PR https://github.com/zephyrproject-rtos/zephyr/pull/60661
I'll rebase this PR once that is merged

gramsay0 committed:
tests: canbus: isotp: conformance: Add tests for CAN-FD support

Allow existing tests to run with CAN-FD.
Add new CAN-FD specific tests.

Signed-off-by: Grant Ramsay <gramsay@enphaseenergy.com>
Signed-off-by: Martin Jäger <martin@libre.solar>

henrikbrixandersen commented on subsys/canbus/isotp/Kconfig:
```diff
@@ -17,6 +17,24 @@ module = ISOTP
 module-str = ISOTP
 source "subsys/logging/Kconfig.template.log_config"
 
+config ISOTP_USE_CAN_FD
+	bool "Use CAN-FD for ISO-TP"
+	depends on CAN_FD_MODE
+	help
+	  Use CAN-FD for ISO-TP.
```
I fought hard to allow for configuring CAN-FD mode at runtime (instead of compile-time) for the CAN controllers (see https://github.com/zephyrproject-rtos/zephyr/issues/45303), so I'm a bit reluctant to add a compile-time option for ISO-TP in this regard. 

Could this be made a run-time configuration with the option to turn it all off via Kconfig (similar as to hiw `CONFIG_CAN_FD_MODE` works)?

gramsay0 replied:
Yes, that could be done, just need to determine how to best add it to the API.
Currently it would need to be passed to `isotp_bind` and the various `isotp_send` functions

@martinjaeger what are your thoughts on this?

martinjaeger replied:
@gramsay0 Sorry for the late response.

I would suggest to add a flag to `struct isotp_msg_id` to indicate if FD frames should be used (similar to `struct can_frame` and `struct can_filter`).

The user will have to call `can_set_mode(dev, CAN_MODE_FD);` manually before using ISO-TP. I think it makes sense that the ISO-TP stack does not automatically switch to CAN-FD, as one could use the same CAN device for multiple different purposes. ISO-TP should use the mode already configured for the device.

gramsay0 replied:
@martinjaeger good idea, the `ide` flag is already in `struct isotp_msg_id` so that makes sense.
I'm on vacation for a couple weeks, feel free to make changes, otherwise I will come back to this later

martinjaeger replied:
I've added a commit to make CAN-FD configurable at runtime. Please have a look @henrikbrixandersen.

henrikbrixandersen reviewed (changes requested):
Is ISO-TP via CAN-FD part of ISO 15765-2?

martinjaeger commented:
> Is ISO-TP via CAN-FD part of ISO 15765-2?

Yes, it's part of the third edition of the spec (from 2016).

martinjaeger commented:
Please ignore the first two commits. The will be removed through rebasing after #61732 is merged.

henrikbrixandersen reviewed (changes requested):
This looks good, @martinjaeger. I'd prefer if the last commit could be squashed into the previous commits to avoid the odd history, though.

martinjaeger commented on tests/subsys/canbus/isotp/conformance/src/main.c:
```diff
@@ -1106,7 +1124,7 @@ void *isotp_conformance_setup(void)
 
 	ret = can_set_mode(can_dev, CAN_MODE_LOOPBACK |
-			   (IS_ENABLED(CONFIG_ISOTP_USE_CAN_FD) ? CAN_MODE_FD : 0));
+			   (IS_ENABLED(CONFIG_CAN_FD_MODE) ? CAN_MODE_FD : 0));
```
In that case `IS_ENABLED(CONFIG_CAN_FD_MODE)` should evaluate to `0` and FD would not be enabled.

Or do you have a case in mind where we'd have one controller with FD capabilities and one without on the same board/SOC, so that we have `CONFIG_CAN_FD_MODE=y` even though one controller doesn't support it?

henrikbrixandersen replied:
`CONFIG_CAN_FD_MODE` can be enabled whether the CAN controller supports CAN-FD or not.

martinjaeger replied:
Yeah, that's exactly how the test is run at the moment, e.g. with `native_posix`. The first case is with default setup `CONFIG_CAN_FD_MODE=n` (even though it would be supported) and the two other cases with extra config `CONFIG_CAN_FD_MODE=y` and some different max data length settings.

Or maybe I misunderstand your question/request?

henrikbrixandersen replied:
The tests need to be able to run on any CAN controller, regardless of `CONFIG_CAN_FD_MODE` being `y` or `n` and whether the CAN controller driver supports CAN-FD or not.

I'd suggest using `can_get_capabilities()` to determine if `CAN_MODE_FD` should be passed to `can_set_mode()` . That does not require any `#ifdef` on `CONFIG_CAN_FD_MODE`.

martinjaeger replied:
It was slightly more complicated than just checking with `can_get_capabilities` and applying the mode accordingly. Now we have the existing `main` test suite, which is skipped if FD is enabled for a controller that does not have FD capabilities. To check that correct errors are returned in such cases, I added another test suite `canfd_notsup.c`.

martinjaeger commented:
> This looks good, @martinjaeger. I'd prefer if the last commit could be squashed into the previous commits to avoid the odd history, though.

I added a new commit so that the authorship of the different changes is clear. But I'm fine with squashing it if @gramsay0 agrees.

dale-herman commented on subsys/canbus/isotp/isotp.c:
```diff
@@ -119,7 +120,8 @@ static inline uint32_t receive_get_sf_length(struct net_buf *buf)
 	struct can_frame frame = {
-		.flags = ctx->tx_addr.ide != 0 ? CAN_FRAME_IDE : 0,
+		.flags = (ctx->tx_addr.ide != 0 ? CAN_FRAME_IDE : 0) |
+			 (IS_ENABLED(ISOTP_USE_CAN_FD) ? CAN_FRAME_FDF | CAN_FRAME_BRS : 0),
```
Should brs (data bitrate switch) be an independent setting dependent on CAN-FD being enabled? This requires the databit rate baud and sampling to be set.

henrikbrixandersen replied:
Good idea. Perhaps we would be better off using a bitmask of options?

martinjaeger replied:
Good point. I've added the `.brs` as as another parameter to the `struct isotp_msg_id`.

I decided against using the existing flags like `CAN_FRAME_FDF` and `CAN_FRAME_BRS`, as we need to apply the configuration both to the filter and the frames. The CAN driver uses different flags for frame and filter, so it would be a bit confusing.

martinjaeger replied:
Ok, that makes sense and I double-checked if we could introduce the flags as part of this PR. However, it would require changes to code unrelated to CAN-FD, as we are currently using two other parameters for the ISO-TP extended addressing and fixed addressing features.

For this reason, I suggest to switch to flags in a follow-up PR.

martinjaeger replied:
Change of mind. Let's get the flags in first: #62255

I'll get back to this PR afterwards.

martinjaeger committed:
tests: canbus: isotp: conformance: add mode_check test

Add a test to check error codes if attempting to use ISO-TP with CAN FD
mode even though the controller supports classical CAN only.

Signed-off-by: Martin Jäger <martin@libre.solar>

henrikbrixandersen commented on tests/subsys/canbus/isotp/conformance/src/canfd_notsup.c:
This logic is reversed to me. We should aim for a generic ISO-TP conformance test suite that is always run, and then have an ISO-TP via CAN-FD test suite that is run if both `CONFIG_CAN_FD_MODE` is enabled _and_ the `zephyr,chosen` CAN controller capabilities includes CAN-FD support, otherwise is should be skipped (or even not compiled-in in the case of `CONFIG_CAN_FD_MODE=n`).

martinjaeger replied:
I agree the naming was not ideal. It is now called `mode_check.c` and it is run for any configuration. The main test suite is also run for any configuration, but skipped if it is configured to use CAN-FD even though it's not supported by the controller.

Note that it is still run if `CONFIG_CAN_FD_MODE=y`, but `CONFIG_TEST_USE_CAN_FD_MODE=n`.

I chose this approach instead of a dedicated CAN-FD test suite because most of the ISO-TP functionality stays the same, CAN-FD only introduces subtile changes in some locations. If we wanted to have one complete test suit for FD and one for classical CAN, we'd end up with lots of code duplication.

martinjaeger commented on subsys/canbus/isotp/Kconfig:
```diff
@@ -17,6 +18,18 @@ module = ISOTP
 
+if CAN_FD_MODE
+
+config ISOTP_TX_DL_MAX
+	int "Transmit data link layer maximum data length"
+	default 64
+	help
+	  The maximum CAN-FD frame data size to send.
+	  Single frames and final frames may still be smaller.
+	  Valid values are 8, 12, 16, 20, 24, 32, 48 and 64.
```
Yes, it is part of the standard:

![image](https://github.com/zephyrproject-rtos/zephyr/assets/17674105/4d9418fb-4228-46e2-8537-480eeb6c890a)

gramsay0 commented:
> > This looks good, @martinjaeger. I'd prefer if the last commit could be squashed into the previous commits to avoid the odd history, though.
> 
> I added a new commit so that the authorship of the different changes is clear. But I'm fine with squashing it if @gramsay0 agrees.

Yep, I'm ok with that.
BTW thanks for carrying this on 👍

alexanderwachter commented on subsys/canbus/isotp/isotp.c:
```diff
@@ -41,6 +42,36 @@ NET_BUF_POOL_VAR_DEFINE(isotp_tx_pool, CONFIG_ISOTP_TX_BUF_COUNT,
 
 static void receive_state_machine(struct isotp_recv_ctx *ctx);
 
+static inline struct can_frame isotp_prepare_frame(struct isotp_msg_id *addr)
```
Not sure if this is going to be efficient. In C++, we can rely on return value optimization (copy elision), but not in C.
I'm not sure if the inline forces to not copy the struct on return.

martinjaeger replied:
I'm absolutely not fluent in assembly. I can at least see that it's not inlined, but kept as a separate function call. Not sure if the function copies the struct on return or not.

<details>
<summary>Disassembly for arm-none-eabi</summary>

```asm
08008008 <prepare_frame>:
{
 8008008:	b538      	push	{r3, r4, r5, lr}
 800800a:	460d      	mov	r5, r1
	struct can_frame frame = {
 800800c:	220d      	movs	r2, #13
 800800e:	2100      	movs	r1, #0
{
 8008010:	4604      	mov	r4, r0
	struct can_frame frame = {
 8008012:	3003      	adds	r0, #3
 8008014:	f001 fb65 	bl	80096e2 <memset>
		.flags = (addr->flags & ISOTP_MSG_IDE) != 0 ? CAN_FRAME_IDE : 0,
 8008018:	79a9      	ldrb	r1, [r5, #6]
		.id = addr->ext_id
 800801a:	682b      	ldr	r3, [r5, #0]
	if ((addr->flags & ISOTP_MSG_FDF) != 0) {
 800801c:	0708      	lsls	r0, r1, #28
		.id = addr->ext_id
 800801e:	f3c3 021c 	ubfx	r2, r3, #0, #29
		.flags = (addr->flags & ISOTP_MSG_IDE) != 0 ? CAN_FRAME_IDE : 0,
 8008022:	f3c1 0380 	ubfx	r3, r1, #2, #1
	if ((addr->flags & ISOTP_MSG_FDF) != 0) {
 8008026:	d505      	bpl.n	8008034 <prepare_frame+0x2c>
		if ((addr->flags & ISOTP_MSG_BRS) != 0) {
 8008028:	06c9      	lsls	r1, r1, #27
		frame.flags |= CAN_FRAME_FDF;
 800802a:	bf54      	ite	pl
 800802c:	f043 0304 	orrpl.w	r3, r3, #4
			frame.flags |= CAN_FRAME_BRS;
 8008030:	f043 030c 	orrmi.w	r3, r3, #12
	return frame;
 8008034:	6821      	ldr	r1, [r4, #0]
 8008036:	7163      	strb	r3, [r4, #5]
 8008038:	f362 011c 	bfi	r1, r2, #0, #29
 800803c:	6021      	str	r1, [r4, #0]
}
 800803e:	4620      	mov	r0, r4
 8008040:	bd38      	pop	{r3, r4, r5, pc}
```
</details>

Even if it takes a copy, it's also a matter of code clarity and reduced duplication. So I'd probably keep it as a function call.

martinjaeger replied:
Done. It's now a void function with pointer argument.

gramsay0 commented on include/zephyr/canbus/isotp.h:
```diff
@@ -167,6 +179,16 @@ struct isotp_msg_id {
+	 * ISO-TP frame data length (TX_DL for TX address or RX_DL for RX address).
+	 *
+	 * Valid values are 8 for classical CAN or 8, 12, 16, 20, 24, 32, 48 and 64 for CAN-FD.
+	 * 0 will be interpreted as 8 for compatibility reasons.
```
I was thinking that doing this at runtime could be useful 👍
Could send lower priority messages as smaller frames to reduce high priority latency.

Maybe also interpret 0 as 64 if CAN-FD is enabled, as that is probably the default expected behavior

alexanderwachter replied:
True.
As far as I remember, 8 is mandatory for classic CAN.

martinjaeger replied:
Yes, I know 8 is mandatory for classical CAN, that's why values < 8 are not in that list. I would rather not ignore it for classical CAN, as you don't want people to choose 6 (or whatever) and implicitly get 8. Default 64 for CAN-FD will come with the next rebase once all the CI issues on main are fixed.

gramsay0 commented on include/zephyr/canbus/isotp.h:
```diff
@@ -147,6 +153,12 @@ extern "C" {
+/** Message uses CAN-FD format (FDF) */
+#define ISOTP_MSG_FDF BIT(3)
+
+/** Message uses CAN-FD Baud Rate Switch (BRS). Only valid in combination with ``CAN_FRAME_FDF``. */
```
Typo: CAN_FRAME_FDF -> ISOTP_MSG_FDF

martinjaeger replied:
Good spot. Fixed.

gramsay0 commented on subsys/canbus/isotp/isotp.c:
```diff
@@ -1163,10 +1236,32 @@ static int send(struct isotp_send_ctx *ctx, const struct device *can_dev,
+			return ISOTP_N_ERROR;
+		}
+		break;
+	}
```
This switch should have a `default:` case that returns an error

gramsay0 commented on tests/subsys/canbus/isotp/conformance/src/main.c:
```diff
@@ -296,14 +370,18 @@ static void prepare_cf_frames(struct frame_desired *frames, size_t frames_cnt,
 		frames[i].data[0] = CF_PCI_BYTE_1 | ((i+1) & 0x0F);
-		frames[i].length = CAN_DL;
+		frames[i].length = TX_DL;
 		memcpy(&des_frames[i].data[1], data_ptr, DATA_SIZE_CF);
```
Not a bug from these changes but `&des_frames[i]` here should be `frames`, and the same thing a few lines below this.
It has no impact on the existing tests, but maybe worth tidying up

martinjaeger replied:
Yeah, you are right. I've added a commit to fix this bug.

gramsay0 reviewed (commented):
Good work!

martinjaeger committed:
tests: canbus: isotp: conformance: fix global variable usage

Even though it was leading to the same result, the function should use
its parameter and not the global variable for desired frames.

Signed-off-by: Martin Jäger <martin@libre.solar>

martinjaeger commented on tests/subsys/canbus/isotp/conformance/src/main.c:
```diff
@@ -949,18 +1021,130 @@ ZTEST(isotp_conformance, test_sender_fc_errors)
+	if (!device_is_ready(can_dev)) {
+		TC_PRINT("CAN device not ready");
+		return false;
+	}
```
Moved back to setup.

martinjaeger commented on tests/subsys/canbus/isotp/conformance/src/main.c:
```diff
@@ -268,7 +310,8 @@ static int add_rx_msgq(uint32_t id, uint32_t mask)
 	struct can_filter filter = {
-		.flags = CAN_FILTER_DATA | ((id > 0x7FF) ? CAN_FILTER_IDE : 0),
+		.flags = CAN_FILTER_DATA | ((id > 0x7FF) ? CAN_FILTER_IDE : 0) |
+			 (IS_ENABLED(CONFIG_CAN_FD_MODE) ? CAN_FILTER_FDF : 0),
```
This is now considered using a test-specific `CONFIG_TEST_USE_CAN_FD_MODE`. `CONFIG_CAN_FD_MODE` can be switched on or off independent of that.

martinjaeger commented:
I hope that all review comments are addressed now and unrelated CI failures are gone.

The ISO-TP test suites and sample were successfully tested on the following boards:
- `native_posix`
- `nucleo_g474re` (CAN-FD support)
- `nucleo_f072rb` (classical CAN only)

henrikbrixandersen commented on tests/subsys/canbus/isotp/conformance/testcase.yaml:
```diff
@@ -1,7 +1,54 @@
+#
+# There are 4 different test cases that have to be considered:
+#
+# | Case # | Controller type        | Selected mode        |  Example board |
+# +--------+------------------------+----------------------+----------------+
+# |    1   | Classical CAN only     | CONFIG_CAN_FD_MODE=n | nucleo_f072    |
+# |    2   | Classical CAN only     | CONFIG_CAN_FD_MODE=y | nucleo_f072    |
+# |    3   | Classical CAN + CAN-FD | CONFIG_CAN_FD_MODE=n | native_posix   |
+# |    4   | Classical CAN + CAN-FD | CONFIG_CAN_FD_MODE=y | native_posix   |
```
These cases significantly increase the time required for running the ISO-TP conformance test suite on real hardware. Can the tests be optimized in time somehow?

Example of time spent running this test suite on five random boards on my desk:
```
INFO    -  6/35 frdm_k64f                 tests/subsys/canbus/isotp/conformance/canbus.isotp.conformance.fd.txdl_64 PASSED (device: 0240000047784e45004e9003d9170022e561000097969900, 8.151s)
INFO    -  7/35 lpcxpresso55s36           tests/subsys/canbus/isotp/conformance/canbus.isotp.conformance.fd.txdl_64 PASSED (device: 001069887263, 14.602s)
INFO    -  8/35 stm32f3_disco             tests/subsys/canbus/isotp/conformance/canbus.isotp.conformance.fd.txdl_64 PASSED (device: 066EFF555052836687113518, 4.123s)
INFO    -  9/35 mimxrt1060_evkb           tests/subsys/canbus/isotp/conformance/canbus.isotp.conformance.fd.txdl_64 PASSED (device: 0229000033f4cd1100000000000000000000000097969905, 33.479s)
INFO    - 10/35 nucleo_g474re             tests/subsys/canbus/isotp/conformance/canbus.isotp.conformance.fd.txdl_64 PASSED (device: 0041001B3438510B34313939, 13.539s)
INFO    - 11/35 frdm_k64f                 tests/subsys/canbus/isotp/conformance/canbus.isotp.conformance.fd.txdl_32 PASSED (device: 0240000047784e45004e9003d9170022e561000097969900, 10.311s)
INFO    - 12/35 lpcxpresso55s36           tests/subsys/canbus/isotp/conformance/canbus.isotp.conformance.fd.txdl_32 PASSED (device: 001069887263, 11.119s)
INFO    - 13/35 stm32f3_disco             tests/subsys/canbus/isotp/conformance/canbus.isotp.conformance.fd.txdl_32 PASSED (device: 066EFF555052836687113518, 3.966s)
INFO    - 14/35 mimxrt1060_evkb           tests/subsys/canbus/isotp/conformance/canbus.isotp.conformance.fd.txdl_32 PASSED (device: 0229000033f4cd1100000000000000000000000097969905, 37.112s)
INFO    - 15/35 nucleo_g474re             tests/subsys/canbus/isotp/conformance/canbus.isotp.conformance.fd.txdl_32 PASSED (device: 0041001B3438510B34313939, 13.679s)
INFO    - 16/35 frdm_k64f                 tests/subsys/canbus/isotp/conformance/canbus.isotp.conformance.fd.unused PASSED (device: 0240000047784e45004e9003d9170022e561000097969900, 16.917s)
INFO    - 17/35 lpcxpresso55s36           tests/subsys/canbus/isotp/conformance/canbus.isotp.conformance.fd.unused PASSED (device: 001069887263, 10.552s)
INFO    - 18/35 stm32f3_disco             tests/subsys/canbus/isotp/conformance/canbus.isotp.conformance.fd.unused PASSED (device: 066EFF555052836687113518, 14.584s)
INFO    - 19/35 mimxrt1060_evkb           tests/subsys/canbus/isotp/conformance/canbus.isotp.conformance.fd.unused PASSED (device: 0229000033f4cd1100000000000000000000000097969905, 33.130s)
INFO    - 20/35 nucleo_g474re             tests/subsys/canbus/isotp/conformance/canbus.isotp.conformance.fd.unused PASSED (device: 0041001B3438510B34313939, 14.217s)
INFO    - 21/35 frdm_k64f                 tests/subsys/canbus/isotp/conformance/canbus.isotp.conformance PASSED (device: 0240000047784e45004e9003d9170022e561000097969900, 17.851s)
INFO    - 22/35 lpcxpresso55s36           tests/subsys/canbus/isotp/conformance/canbus.isotp.conformance PASSED (device: 001069887263, 11.913s)
INFO    - 23/35 mimxrt1060_evkb           tests/subsys/canbus/isotp/conformance/canbus.isotp.conformance PASSED (device: 0229000033f4cd1100000000000000000000000097969905, 31.035s)
INFO    - 24/35 stm32f3_disco             tests/subsys/canbus/isotp/conformance/canbus.isotp.conformance PASSED (device: 066EFF555052836687113518, 14.670s)
INFO    - 25/35 nucleo_g474re             tests/subsys/canbus/isotp/conformance/canbus.isotp.conformance PASSED (device: 0041001B3438510B34313939, 14.054s)
```

Do we really need to test with different TX_DLs on all boards or could a simple unit-like test (e.g. on POSIX) be sufficient for the different configurations? I mean, these tests validate the protocol implementation, not as much as the protocol interaction with the particular CAN controller?

Another option could be to tag some of these tests as "slow".

martinjaeger replied:
There will be some more upcoming ISO-TP updates. Let's leave the optimization of test time for a future PR.

henrikbrixandersen reviewed (dismissed):
Looks good. A couple of minor observations:

dale-herman reviewed (approved):
Looks good.

martinjaeger commented on tests/subsys/canbus/isotp/conformance/src/mode_check.c:
```diff
@@ -0,0 +1,91 @@
+	return NULL;
+}
+
+ZTEST_SUITE(isotp_canfd_notsup, NULL, isotp_canfd_notup_setup, NULL, NULL, NULL);
```
Fixed.

martinjaeger commented:
Fixed the typo in the test naming identified by @henrikbrixandersen. Please re-approve once again, @alexanderwachter . Thanks!

martinjaeger reviewed (dismissed):
Approval and thanks for the parts of the work @gramsay0 and @dale-herman have done. Also thanks for all reviews.

alexanderwachter reviewed (approved):
Thanks Martin!
